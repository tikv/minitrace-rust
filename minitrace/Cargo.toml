[package]
name = "minitrace"
version = "0.6.6"
authors = ["The TiKV Project Authors"]
license = "Apache-2.0"
edition = "2021"
description = "A high-performance timeline tracing library for Rust"
homepage = "https://github.com/tikv/minitrace-rust"
repository = "https://github.com/tikv/minitrace-rust"
documentation = "https://docs.rs/minitrace"
readme = "../README.md"
categories = ["development-tools::debugging"]
keywords = ["tracing", "span", "datadog", "jaeger", "opentelemetry"]

[features]
enable = []

[dependencies]
minitrace-macro = { version = "0.6.6", path = "../minitrace-macro" }
minstant = "0.1.7"
parking_lot = "0.12"
pin-project = "1"
# TODO: Remove once_cell once #![feature(once_cell)] is stabilized
once_cell = "1"
rand = "0.8"
rtrb = "0.2"

[dev-dependencies]
# The procedural macro `trace` only supports async-trait higher than 0.1.52
async-trait = "0.1.52"
criterion = { version = "0.4", features = ["html_reports"] }
crossbeam = "0.8"
env_logger = "0.10"
futures = "0.3"
futures-timer = "3"
log = "0.4"
logcall = "0.1.4"
minitrace = { path = ".", features = ["enable"] }
minitrace-datadog = { version = "0.6.6", path = "../minitrace-datadog" }
minitrace-jaeger = { version = "0.6.6", path = "../minitrace-jaeger" }
minitrace-opentelemetry = { version = "0.6.6", path = "../minitrace-opentelemetry" }
mockall = "0.11"
once_cell = "1"
opentelemetry-otlp = { version = "0.16", features = ["trace"] }
opentelemetry = { version = "0.23", features = ["trace"] }
opentelemetry_sdk = { version = "0.23", features = ["trace"] }

rand = "0.8"
rustracing = "0.6"
serial_test = "2"
test-harness = "0.1.1"
tokio = { version = "1", features = ["rt", "time", "macros", "rt-multi-thread"] }
tracing = "0.1"
tracing-core = "0.1"
tracing-opentelemetry = "0.24"
tracing-subscriber = "0.3"

[[bench]]
name = "trace"
harness = false

[[bench]]
name = "compare"
harness = false

[[bench]]
name = "spsc"
harness = false

[[bench]]
name = "object_pool"
harness = false
