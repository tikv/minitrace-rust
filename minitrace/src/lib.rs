// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.

//! A high-performance, ergonomic timeline tracing library for Rust.
//!
//! ## Span
//!
//!   A [`SpanRecord`] represents an individual unit of work. It contains:
//!   - An operation name
//!   - A start timestamp and duration
//!   - A set of key-value properties
//!   - A reference to a parent `Span`
//!
//!   To record such a span record, we create a [`Span`] to start clocking and drop it to stop recording.
//!
//!   A new `Span` can be started via [`Span::root()`] and [`Span::enter_with_parent()`]. `Span::enter_with_parent()`
//!   will start a child span to a given parent span.
//!
//!   `Span` is thread-safe and can be sent across threads.
//!
//!   ```
//!   use minitrace::prelude::*;
//!   use futures::executor::block_on;
//!
//!   let (root, collector) = Span::root("root");
//!
//!   {
//!       let _child_span = Span::enter_with_parent("a child span", &root);
//!       // some work
//!   }
//!
//!   drop(root);
//!   let records: Vec<SpanRecord> = block_on(collector.collect());
//!
//!   println!("{records:#?}");
//!   // [
//!   //     SpanRecord {
//!   //         id: 1,
//!   //         parent_id: 0,
//!   //         begin_unix_time_ns: 1642166520139678013,
//!   //         duration_ns: 16008,
//!   //         name: "root",
//!   //         properties: [],
//!   //         events: [],
//!   //     },
//!   //     SpanRecord {
//!   //         id: 2,
//!   //         parent_id: 1,
//!   //         begin_unix_time_ns: 1642166520139692070,
//!   //         duration_ns: 634,
//!   //         name: "a child span",
//!   //         properties: [],
//!   //         events: [],
//!   //     },
//!   // ]
//!   ```
//!
//!
//! ## Local Span
//!
//!   A `Span` can be optimized into [`LocalSpan`], if the span is not supposed to be sent to other threads,
//!   which can greatly reduce the overhead.
//!
//!   Before starting a `LocalSpan`, a scope of parent span should be set using [`Span::set_local_parent()`].
//!   Use [`LocalSpan::enter_with_local_parent()`] to start a `LocalSpan`, and then, it will become the new local parent.
//!
//!   If no local parent is set, the `enter_with_local_parent()` will do nothing.
//!
//!   ```
//!   use minitrace::prelude::*;
//!   use futures::executor::block_on;
//!
//!   let (root, collector) = Span::root("root");
//!
//!   {
//!       let _guard = root.set_local_parent();
//!
//!       // The parent of this span is `root`.
//!       let _span1 = LocalSpan::enter_with_local_parent("a child span");
//!
//!       foo();
//!   }
//!
//!   fn foo() {
//!       // The parent of this span is `span1`.
//!       let _span2 = LocalSpan::enter_with_local_parent("a child span of child span");
//!   }
//!
//!   drop(root);
//!   let records: Vec<SpanRecord> = block_on(collector.collect());
//!
//!   println!("{records:#?}");
//!   // [
//!   //     SpanRecord {
//!   //         id: 1,
//!   //         parent_id: 0,
//!   //         begin_unix_time_ns: 1643101008017429580,
//!   //         duration_ns: 64132,
//!   //         name: "root",
//!   //         properties: [],
//!   //         events: [],
//!   //     },
//!   //     SpanRecord {
//!   //         id: 2,
//!   //         parent_id: 1,
//!   //         begin_unix_time_ns: 1643101008017486383,
//!   //         duration_ns: 4150,
//!   //         name: "a child span",
//!   //         properties: [],
//!   //         events: [],
//!   //     },
//!   //     SpanRecord {
//!   //         id: 3,
//!   //         parent_id: 2,
//!   //         begin_unix_time_ns: 1643101008017488703,
//!   //         duration_ns: 1318,
//!   //         name: "a child span of child span",
//!   //         properties: [],
//!   //         events: [],
//!   //     },
//!   // ]
//!   ```
//!
//!
//! ## Property
//!
//!   Property is an arbitrary custom kev-value pair associated to a span.
//!
//!   ```
//!   use minitrace::prelude::*;
//!   use futures::executor::block_on;
//!
//!   let (mut root, collector) = Span::root("root");
//!   root.add_property(|| ("key", "value".to_owned()));
//!
//!   {
//!       let _guard = root.set_local_parent();
//!
//!       let mut span1 = LocalSpan::enter_with_local_parent("a child span");
//!       span1.add_property(|| ("key", "value".to_owned()));
//!   }
//!
//!   drop(root);
//!   let records: Vec<SpanRecord> = block_on(collector.collect());
//!
//!   println!("{records:#?}");
//!   // [
//!   //     SpanRecord {
//!   //         id: 1,
//!   //         parent_id: 0,
//!   //         begin_unix_time_ns: 1642166791041022255,
//!   //         duration_ns: 121705,
//!   //         name: "root",
//!   //         properties: [
//!   //             (
//!   //                 "key",
//!   //                 "value",
//!   //             ),
//!   //         ],
//!   //         events: [],
//!   //     },
//!   //     SpanRecord {
//!   //         id: 2,
//!   //         parent_id: 1,
//!   //         begin_unix_time_ns: 1642166791041132550,
//!   //         duration_ns: 7724,
//!   //         name: "a child span",
//!   //         properties: [
//!   //             (
//!   //                 "key",
//!   //                 "value",
//!   //             ),
//!   //         ],
//!   //         events: [],
//!   //     },
//!   // ]
//!   ```
//!
//!
//! ## Event
//!
//!   [`Event`] represent single points in time where something occurred during the execution of a program.
//!   An `Event` can be seen as a log record attached to a span.
//!
//!   ```
//!   use minitrace::prelude::*;
//!   use futures::executor::block_on;
//!
//!   let (mut root, collector) = Span::root("root");
//!
//!   Event::add_to_parent("event in root", &root, || []);
//!
//!   {
//!       let _guard = root.set_local_parent();
//!       let mut span1 = LocalSpan::enter_with_local_parent("a child span");
//!
//!       Event::add_to_local_parent("event in span1", || [("key", "value".to_owned())]);
//!   }
//!
//!   drop(root);
//!   let records: Vec<SpanRecord> = block_on(collector.collect());
//!
//!   println!("{records:#?}");
//!   // [
//!   //     SpanRecord {
//!   //         id: 1,
//!   //         parent_id: 0,
//!   //         begin_unix_time_ns: 1689321940550848459,
//!   //         duration_ns: 25708,
//!   //         name: "root",
//!   //         properties: [],
//!   //         events: [
//!   //             EventRecord {
//!   //                 name: "event in root",
//!   //                 timestamp_unix_ns: 1689321940550870667,
//!   //                 properties: [],
//!   //             },
//!   //         ],
//!   //     },
//!   //     SpanRecord {
//!   //         id: 3,
//!   //         parent_id: 1,
//!   //         begin_unix_time_ns: 1689321940550874167,
//!   //         duration_ns: 0,
//!   //         name: "span1",
//!   //         properties: [],
//!   //         events: [
//!   //             EventRecord {
//!   //                 name: "event in span1",
//!   //                 timestamp_unix_ns: 1689321940550874167,
//!   //                 properties: [
//!   //                     (
//!   //                         "key",
//!   //                         "value",
//!   //                     ),
//!   //                 ],
//!   //             },
//!   //         ],
//!   //     },
//!   // ]
//!   ```
//!
//!
//! ## Macro
//!
//!   An attribute-macro [`trace`] can help get rid of boilerplate. The macro always requires a local
//!   parent in the context, otherwise, no span will be recorded.
//!
//!   ```
//!   use minitrace::prelude::*;
//!   use futures::executor::block_on;
//!
//!   #[trace]
//!   fn do_something(i: u64) {
//!       std::thread::sleep(std::time::Duration::from_millis(i));
//!   }
//!
//!   #[trace]
//!   async fn do_something_async(i: u64) {
//!       futures_timer::Delay::new(std::time::Duration::from_millis(i)).await;
//!   }
//!
//!   let (root, collector) = Span::root("root");
//!
//!   {
//!       let _g = root.set_local_parent();
//!       do_something(100);
//!       block_on(do_something_async(100));
//!   }
//!
//!   drop(root);
//!   let records: Vec<SpanRecord> = block_on(collector.collect());
//!
//!   println!("{records:#?}");
//!   // [
//!   //     SpanRecord {
//!   //         id: 1,
//!   //         parent_id: 0,
//!   //         begin_unix_time_ns: 1642167988459480418,
//!   //         duration_ns: 200741472,
//!   //         name: "root",
//!   //         properties: [],
//!   //     },
//!   //     SpanRecord {
//!   //         id: 2,
//!   //         parent_id: 1,
//!   //         begin_unix_time_ns: 1642167988459571971,
//!   //         duration_ns: 100084126,
//!   //         name: "do_something",
//!   //         properties: [],
//!   //     },
//!   //     SpanRecord {
//!   //         id: 3,
//!   //         parent_id: 1,
//!   //         begin_unix_time_ns: 1642167988559887219,
//!   //         duration_ns: 100306947,
//!   //         name: "do_something_async",
//!   //         properties: [],
//!   //     },
//!   // ]
//!   ```
//!
//! [`Span`]: crate::Span
//! [`LocalSpan`]: crate::local::LocalSpan
//! [`SpanRecord`]: crate::collector::SpanRecord
//! [`FutureExt`]: crate::future::FutureExt
//! [`trace`]: crate::trace
//! [`LocalCollector`]: crate::local::LocalCollector
//! [`Span::root()`]: crate::Span::root
//! [`Span::enter_with_parent()`]: crate::Span::enter_with_parent
//! [`Span::set_local_parent()`]: crate::Span::set_local_parent
//! [`LocalSpan::enter_with_local_parent()`]: crate::local::LocalSpan::enter_with_local_parent

// Suppress a false-positive lint from clippy
// TODO: remove me once https://github.com/rust-lang/rust-clippy/issues/11076 is released
#![allow(unknown_lints)]
#![allow(clippy::arc_with_non_send_sync)]

pub mod collector;
mod event;
pub mod future;
pub mod local;
mod span;
#[doc(hidden)]
pub mod util;

/// An attribute-macro to help get rid of boilerplate.
///
/// The span name is the function name by default. It can be customized by passing a string literal.
///
/// [`trace`] always require an local parent in the context. Make sure that the caller
/// is within the scope of [`Span::set_local_parent()`].
///
/// # Examples
///
/// ```
/// use minitrace::prelude::*;
///
/// #[trace]
/// fn foo() {
///     // some work
/// }
///
/// #[trace]
/// async fn bar() {
///     // some work
/// }
///
/// #[trace("qux", enter_on_poll = true)]
/// async fn qux() {
///     // some work
/// }
/// ```
///
/// The examples above will be translated into:
///
/// ```
/// # use minitrace::prelude::*;
/// # use minitrace::local::LocalSpan;
/// fn foo() {
///     let __guard = LocalSpan::enter_with_local_parent("foo");
///     // some work
/// }
///
/// fn bar() -> impl core::future::Future<Output = ()> {
///     async {
///         // some work
///     }
///     .in_span(Span::enter_with_local_parent("bar"))
/// }
///
/// fn qux() -> impl core::future::Future<Output = ()> {
///     async {
///         // some work
///     }
///     .enter_on_poll("qux")
/// }
/// ```
///
/// [`in_span()`]: crate::future::FutureExt::in_span
pub use minitrace_macro::trace;

pub use crate::event::Event;
pub use crate::span::Span;

pub mod prelude {
    //! A "prelude" for crates using the `minitrace` crate.
    #[doc(no_inline)]
    pub use crate::collector::CollectArgs;
    #[doc(no_inline)]
    pub use crate::collector::Collector;
    #[doc(no_inline)]
    pub use crate::collector::SpanRecord;
    #[doc(no_inline)]
    pub use crate::event::Event;
    #[doc(no_inline)]
    pub use crate::future::FutureExt as _;
    #[doc(no_inline)]
    pub use crate::local::LocalSpan;
    #[doc(no_inline)]
    pub use crate::span::Span;
    #[doc(no_inline)]
    pub use crate::trace;
}
