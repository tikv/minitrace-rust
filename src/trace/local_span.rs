// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.

use crate::LocalSpanGuard;

pub struct LocalSpan;

impl LocalSpan {
    pub fn enter(event: &'static str) -> LocalSpanGuard {
        LocalSpanGuard::new(event)
    }
}
