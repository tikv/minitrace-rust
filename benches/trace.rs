// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.

use criterion::{black_box, criterion_group, criterion_main, Criterion};
use minitrace::local::LocalCollector;
use minitrace::prelude::*;
use minitrace_macro::trace;

fn dummy_iter(i: usize) {
    #[trace("")]
    fn dummy() {}

    for _ in 0..i {
        dummy();
    }
}

#[trace("")]
fn dummy_rec(i: usize) {
    if i > 1 {
        dummy_rec(i - 1);
    }
}

fn trace_wide_raw_bench(c: &mut Criterion) {
    let mut group = c.benchmark_group("trace_wide_raw");

    for len in &[1, 10, 100, 1000, 10000] {
        group.bench_function(len.to_string(), |b| {
            b.iter(|| {
                let local_collector = LocalCollector::start();
                dummy_iter(*len);
                local_collector.collect()
            })
        });
    }

    group.finish();
}

fn trace_wide_bench(c: &mut Criterion) {
    let mut group = c.benchmark_group("trace_wide");

    for len in &[1, 10, 100, 1000, 10000] {
        group.bench_function(len.to_string(), |b| {
            b.iter(|| {
                {
                    let (root_span, collector) = Span::root("root");
                    let _sg = root_span.set_local_parent();
                    dummy_iter(*len - 1);
                    collector
                }
                .collect()
            })
        });
    }

    group.finish();
}

fn trace_deep_raw_bench(c: &mut Criterion) {
    let mut group = c.benchmark_group("trace_deep_raw");

    for len in &[1, 10, 100, 1000] {
        group.bench_function(len.to_string(), |b| {
            b.iter(|| {
                let local_collector = LocalCollector::start();
                dummy_rec(*len);
                local_collector.collect()
            })
        });
    }

    group.finish();
}

fn trace_deep_bench(c: &mut Criterion) {
    let mut group = c.benchmark_group("trace_deep");

    for len in &[1, 10, 100, 1000] {
        group.bench_function(len.to_string(), |b| {
            b.iter(|| {
                {
                    let (root_span, collector) = Span::root("root");
                    let _sg = root_span.set_local_parent();
                    dummy_rec(*len - 1);
                    collector
                }
                .collect()
            })
        });
    }

    group.finish();
}

fn trace_future_bench(c: &mut Criterion) {
    async fn f(i: u32) {
        for _ in 0..i - 1 {
            async {}.enter_on_poll(black_box("")).await
        }
    }

    let mut group = c.benchmark_group("trace_future");

    for len in &[1, 10, 100, 1000, 10000] {
        group.bench_function(len.to_string(), |b| {
            b.iter(|| {
                {
                    let (root_span, collector) = Span::root("root");
                    let _ = futures::executor::block_on(f(*len).in_span(root_span));
                    collector
                }
                .collect()
            })
        });
    }

    group.finish();
}

criterion_group!(
    benches,
    trace_wide_raw_bench,
    trace_wide_bench,
    trace_deep_raw_bench,
    trace_deep_bench,
    trace_future_bench
);
criterion_main!(benches);
