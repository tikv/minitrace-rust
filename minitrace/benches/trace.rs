// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.

use criterion::{black_box, criterion_group, criterion_main, Criterion};
use minitrace::local::LocalCollector;
use minitrace::prelude::*;

fn dummy_iter(i: usize) {
    #[trace]
    fn dummy() {}

    for _ in 0..i {
        dummy();
    }
}

#[trace]
fn dummy_rec(i: usize) {
    if i > 1 {
        dummy_rec(i - 1);
    }
}

fn bench_trace_wide_raw(c: &mut Criterion) {
    let mut group = c.benchmark_group("trace_wide_raw");

    for len in &[1, 10, 100, 1000, 10000] {
        group.bench_function(len.to_string(), |b| {
            b.iter(|| {
                let local_collector = LocalCollector::start();
                dummy_iter(*len);
                local_collector.collect()
            })
        });
    }

    group.finish();
}

fn bench_trace_wide(c: &mut Criterion) {
    let mut group = c.benchmark_group("trace_wide");

    for len in &[1, 10, 100, 1000, 10000] {
        group.bench_function(format!("with-collect-{}", len), |b| {
            b.iter(|| {
                {
                    let (root_span, collector) = Span::root("root");
                    let _sg = root_span.set_local_parent();
                    dummy_iter(*len - 1);
                    collector
                }
                .collect()
            })
        });
        group.bench_function(format!("without-collect-{}", len), |b| {
            b.iter(|| {
                let (root_span, _) = Span::root("root");
                let _sg = root_span.set_local_parent();
                dummy_iter(*len - 1);
            })
        });
    }

    group.finish();
}

fn bench_trace_deep_raw(c: &mut Criterion) {
    let mut group = c.benchmark_group("trace_deep_raw");

    for len in &[1, 10, 100, 1000] {
        group.bench_function(len.to_string(), |b| {
            b.iter(|| {
                let local_collector = LocalCollector::start();
                dummy_rec(*len);
                local_collector.collect()
            })
        });
    }

    group.finish();
}

fn bench_trace_deep(c: &mut Criterion) {
    let mut group = c.benchmark_group("trace_deep");

    for len in &[1, 10, 100, 1000] {
        group.bench_function(format!("with-collect-{}", len), |b| {
            b.iter(|| {
                {
                    let (root_span, collector) = Span::root("root");
                    let _sg = root_span.set_local_parent();
                    dummy_rec(*len - 1);
                    collector
                }
                .collect()
            })
        });
        group.bench_function(format!("without-collect-{}", len), |b| {
            b.iter(|| {
                let (root_span, _) = Span::root("root");
                let _sg = root_span.set_local_parent();
                dummy_rec(*len - 1);
            })
        });
    }

    group.finish();
}

fn bench_trace_future(c: &mut Criterion) {
    async fn f(i: u32) {
        for _ in 0..i - 1 {
            async {}.enter_on_poll(black_box("")).await
        }
    }

    let mut group = c.benchmark_group("trace_future");

    for len in &[1, 10, 100, 1000, 10000] {
        group.bench_function(len.to_string(), |b| {
            b.iter(|| {
                {
                    let (root_span, collector) = Span::root("root");
                    futures::executor::block_on(f(*len).in_span(root_span));
                    collector
                }
                .collect()
            })
        });
    }

    group.finish();
}

criterion_group!(
    benches,
    bench_trace_wide_raw,
    bench_trace_wide,
    bench_trace_deep_raw,
    bench_trace_deep,
    bench_trace_future
);
criterion_main!(benches);
