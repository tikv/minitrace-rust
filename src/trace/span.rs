// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.

use std::iter;
use std::sync::atomic::AtomicBool;
use std::sync::Arc;

use crate::local::local_collector::LocalSpans;
use crate::span::RawSpan;
use crate::span::{DefaultClock, DefaultIdGenerator, SpanId};
use crate::trace::acquirer::{Acquirer, SpanCollection};
use crate::Collector;

#[must_use]
#[derive(Debug)]
pub struct Span {
    pub(crate) inner: Option<SpanInner>,
}

#[derive(Debug)]
pub(crate) struct SpanInner {
    pub(crate) span_id: SpanId,

    // Report `RawSpan` to `Acquirer` when `SpanInner` is dropping
    pub(crate) to_report: Vec<(RawSpan, Acquirer)>,
}

impl Span {
    #[inline]
    pub(crate) fn new<'a>(
        acquirers: impl Iterator<Item = (SpanId, &'a Acquirer)>,
        event: String,
    ) -> Self {
        let span_id = DefaultIdGenerator::next_id();
        let now = DefaultClock::now();

        let mut to_report = Vec::new();
        for (parent_span_id, acq) in acquirers {
            if !acq.is_shutdown() {
                to_report.push((
                    RawSpan::begin_with(span_id, parent_span_id, now, event.clone()),
                    acq.clone(),
                ))
            }
        }

        if to_report.is_empty() {
            Self { inner: None }
        } else {
            Self {
                inner: Some(SpanInner { span_id, to_report }),
            }
        }
    }

    pub fn root(event: String) -> (Self, Collector) {
        let (tx, rx) = crossbeam::channel::unbounded();
        let closed = Arc::new(AtomicBool::new(false));
        let acquirer = Acquirer::new(Arc::new(tx), closed.clone());
        let span = Self::new(iter::once((SpanId::new(0), &acquirer)), event);
        let collector = Collector::new(rx, closed);
        (span, collector)
    }

    #[inline]
    pub fn empty() -> Self {
        Self { inner: None }
    }

    #[inline]
    pub fn is_empty(&self) -> bool {
        self.inner.is_none()
    }

    #[inline]
    pub fn from_parent(event: String, span: &Span) -> Self {
        Self::from_parents(event, iter::once(span))
    }

    #[inline]
    pub fn from_parents<'a>(
        event: String,
        spans: impl IntoIterator<Item = &'a Span>,
    ) -> Self {
        Self::new(
            spans
                .into_iter()
                .filter_map(|span| span.inner.as_ref())
                .flat_map(|inner| {
                    inner
                        .to_report
                        .iter()
                        .map(move |(_, acq)| (inner.span_id, acq))
                }),
            event,
        )
    }

    #[inline]
    pub fn mount_local_spans(&self, local_spans: Arc<LocalSpans>) {
        if let Some(inner) = &self.inner {
            for (_, acq) in &inner.to_report {
                acq.submit(SpanCollection::LocalSpans {
                    local_spans: local_spans.clone(),
                    parent_id_of_root: inner.span_id,
                })
            }
        }
    }
}

impl Drop for SpanInner {
    fn drop(&mut self) {
        let now = DefaultClock::now();
        for (mut span, collector) in self.to_report.drain(..) {
            span.end_with(now);
            collector.submit(SpanCollection::Span(span))
        }
    }
}
