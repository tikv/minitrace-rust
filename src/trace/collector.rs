// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.

use crossbeam::channel::Receiver;

use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Duration;

use crate::span::Span;
use crate::span::{Anchor, DefaultClock};
use crate::trace::acquirer::SpanCollection;

pub struct Collector {
    receiver: Receiver<SpanCollection>,
    closed: Arc<AtomicBool>,
}

impl Collector {
    pub(crate) fn new(receiver: Receiver<SpanCollection>, closed: Arc<AtomicBool>) -> Self {
        Collector { receiver, closed }
    }

    pub fn collect(self) -> Vec<Span> {
        self.collect_with_args(CollectArgs {
            sync: false,
            duration_threshold: None,
        })
    }

    /// Collects spans from traced routines.
    ///
    /// If passing `duration_threshold`, all spans will be reserved only when duration of the root
    /// span exceeds `duration_threshold`, otherwise only one span, the root span, will be returned.
    pub fn collect_with_args(
        self,
        CollectArgs {
            sync,
            duration_threshold,
        }: CollectArgs,
    ) -> Vec<Span> {
        let span_collections: Vec<_> = if sync {
            self.receiver.iter().collect()
        } else {
            self.receiver.try_iter().collect()
        };
        self.closed.store(true, Ordering::SeqCst);

        let anchor = DefaultClock::anchor();
        if let Some(duration) = duration_threshold {
            // find the root span and check its duration
            if let Some(root_span) = span_collections.iter().find_map(|s| match s {
                SpanCollection::Span(s) if s.parent_id.0 == 0 => Some(s),
                _ => None,
            }) {
                let root_span = root_span.clone().into_span(anchor);
                if root_span.duration_ns < duration.as_nanos() as _ {
                    return vec![root_span];
                }
            }
        }

        Self::amend(span_collections, anchor)
    }
}

impl Collector {
    #[inline]
    fn amend(span_collections: Vec<SpanCollection>, anchor: Anchor) -> Vec<Span> {
        let capacity = span_collections
            .iter()
            .map(|sc| match sc {
                SpanCollection::LocalSpans {
                    local_spans: raw_spans,
                    ..
                } => raw_spans.spans.len(),
                SpanCollection::Span(_) => 1,
            })
            .sum();

        let mut spans = Vec::with_capacity(capacity);

        for span_collection in span_collections {
            match span_collection {
                SpanCollection::LocalSpans {
                    local_spans: raw_spans,
                    parent_id_of_root: span_id,
                } => {
                    for span in &raw_spans.spans {
                        let begin_unix_time_ns =
                            DefaultClock::cycle_to_unix_time_ns(span.begin_cycle, anchor);
                        let end_unix_time_ns = if span.end_cycle.is_zero() {
                            DefaultClock::cycle_to_unix_time_ns(raw_spans.end_time, anchor)
                        } else {
                            DefaultClock::cycle_to_unix_time_ns(span.end_cycle, anchor)
                        };
                        let parent_id = if span.parent_id.0 == 0 {
                            span_id.0
                        } else {
                            span.parent_id.0
                        };
                        spans.push(Span {
                            id: span.id.0,
                            parent_id,
                            begin_unix_time_ns,
                            duration_ns: end_unix_time_ns - begin_unix_time_ns,
                            event: span.event.clone(),
                            properties: span.properties.clone(),
                        });
                    }
                }
                SpanCollection::Span(span) => spans.push(span.into_span(anchor)),
            }
        }

        spans
    }
}

#[derive(Default, Debug)]
pub struct CollectArgs {
    sync: bool,
    duration_threshold: Option<Duration>,
}

impl CollectArgs {
    pub fn sync(self, sync: bool) -> Self {
        Self { sync, ..self }
    }

    pub fn duration_threshold(self, duration_threshold: Duration) -> Self {
        Self {
            duration_threshold: Some(duration_threshold),
            ..self
        }
    }
}
