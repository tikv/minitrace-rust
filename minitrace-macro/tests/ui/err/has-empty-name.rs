use minitrace::trace;

#[trace(name = "")]
fn f() {}

fn main() {}
