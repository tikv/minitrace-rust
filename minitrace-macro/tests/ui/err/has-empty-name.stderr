error: `name` can not be empty
 --> tests/ui/err/has-empty-name.rs:3:1
  |
3 | #[trace(name = "")]
  | ^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the attribute macro `trace` (in Nightly builds, run with -Z macro-backtrace for more info)
